src/storage/exec/UpdateNode.h

```diff
@@ -167,57 +167,64 @@ class UpdateTagNode : public UpdateNode<VertexID> {
-                        this->exeResult_ = this->insertTagProps(partId, vId);
-                    } else if (this->reader_) {
-                        this->key_ = filterNode_->key().str();
-                        this->exeResult_ = this->collTagProp(vId);
-                    } else {
-                        this->exeResult_ = kvstore::ResultCode::ERR_KEY_NOT_FOUND;
-                    }
+        if (ret != kvstore::ResultCode::SUCCEEDED) {
+            return ret;
+        }
 
-                    if (this->exeResult_ != kvstore::ResultCode::SUCCEEDED) {
-                        return folly::none;
-                    }
-                    return this->updateAndWriteBack(partId, vId);
-                } else {
-                    // if tagnode/edgenode error
-                    return folly::none;
-                }
-            },
-            [&ret, &baton, this] (kvstore::ResultCode code) {
-                if (code == kvstore::ResultCode::ERR_ATOMIC_OP_FAILED &&
-                    this->exeResult_ != kvstore::ResultCode::SUCCEEDED) {
-                    ret = this->exeResult_;
-                } else {
-                    ret = code;
-                }
-                baton.post();
-            });
+        if (this->planContext_->resultStat_ == ResultStatus::ILLEGAL_DATA) {
+            return kvstore::ResultCode::ERR_INVALID_DATA;
+        } else if (this->planContext_->resultStat_ ==  ResultStatus::FILTER_OUT) {
+            return kvstore::ResultCode::ERR_RESULT_FILTERED;
+        }
+
+        if (filterNode_->valid()) {
+            this->reader_ = filterNode_->reader();
+        }
+        // reset StorageExpressionContext reader_, because it contains old value
+        this->expCtx_->reset();
+
+        if (!this->reader_ && this->insertable_) {
+            ret = this->insertTagProps(partId, vId);
+        } else if (this->reader_) {
+            this->key_ = filterNode_->key().str();
+            ret = this->collTagProp(vId);
+        } else {
+            ret = kvstore::ResultCode::ERR_KEY_NOT_FOUND;
+        }
+
+        if (ret != kvstore::ResultCode::SUCCEEDED) {
+            return ret;
+        }
+
+        auto batch = this->updateAndWriteBack(partId, vId);
+        if (batch == folly::none) {
+            return kvstore::ResultCode::ERR_INVALID_DATA;
+        }
+
+        std::vector<VMLI> dummyLock = {std::make_tuple(planContext_->spaceId_,
+                                                       partId, tagId_, vId)};
+        nebula::MemoryLockGuard<VMLI> lg(planContext_->env_->verticesML_.get(), dummyLock);
+        if (!lg) {
+            auto conflict = lg.conflictKey();
+            LOG(ERROR) << "vertex conflict "
+                       << std::get<0>(conflict) << ":"
+                       << std::get<1>(conflict) << ":"
+                       << std::get<2>(conflict) << ":"
+                       << std::get<3>(conflict);
+            return kvstore::ResultCode::ERR_CONSENSUS_ERROR;
```

Comment:
> Better return another error code? Storage client could retry if we add retry later.

I think here is no need to retry always. we need to let the user know where the conflict is, user retries at the business layer. 
Also, if we need to retry, the best way is just skip this conflict row and report warning, not need return error code. WDYT?

Reply:
Why do you think we should not retry when conflicts? 
I mean if we add retry later, when conflict, storage client should take some special handling just like leader change?

Reply:
> Why do you think we should not retry when conflicts?
> I mean if we add retry later, when conflict, storage client should take some special handling just like leader change?
Because we don't know what the final data is if there's a retry
for example : insert vertex , the vertex id is 1
retry : 
first time
thread 1 : insert vertex 1 (conflicted)
thread 2 : insert vertex 1 (conflicted)
second time
thread 1 : insert vertex 1 (success | conflicted then retry)
thread 2 : insert vertex 1 (success  | conflicted then retry)
if conflicted , then retry ... it was successful in the end
the final data is thread 1's data, right? Or is it thread 2?

If retry is supported, there is a better way：
first time
thread 1 : insert vertex 1 (conflicted, skip the row in mem lock and report warning, then write data immediate)
thread 2 : insert vertex 1 (conflicted, skip the row in mem lock and report warning, then write data immediate)
it was successful same with retry.

Reply:
Yeah, retry logic make sense. Anyway, the result code need to change, `ERR_CONSENSUS_ERROR` is used in other case.

Reply:
> Yeah, retry logic make sense. Anyway, the result code need to change, `ERR_CONSENSUS_ERROR` is used in other case.

make sense, what's about `ERR_CONFLICT_DATA`?

Reply:
👌